File: compiler/testData/diagnostics/tests/dataFlowInfoTraversal/ArrayAccess.kt
fun bar(x: Int): Int = x + 1

fun foo() {
    val x: Int? = null
    val a = Array<Int>(3, {0})

    if (x != null) bar(a[<!DEBUG_INFO_SMARTCAST!>x<!>]) else bar(a[<!DEBUG_INFO_CONSTANT, TYPE_MISMATCH!>x<!>])
    bar(a[if (x == null) 0 else <!DEBUG_INFO_SMARTCAST!>x<!>])
    bar(a[<!TYPE_MISMATCH!>x<!>])
    
    "123"[<!TYPE_MISMATCH!>x<!>];
    if (x != null) "123"[<!DEBUG_INFO_SMARTCAST!>x<!>];
}

File: js/js.translator/testData/dce/arrayAccess.js
(function(_) {
    var x = {};
    function foo() {
        return x["bar"]();
    }
    x.bar = function() {
        return "bar";
    };
    x.baz = function() {
        return "baz";
    };
    _.foo = foo;
})(module.exports);

// REQUEST_REACHABLE: main.foo
// ASSERT_REACHABLE: x.bar
// ASSERT_UNREACHABLE: x.baz